GEM and ME0 realistic digitizer updates

dildick opened the pull request:
This pull request includes updates for the ME0 digtizer

Below is a list of things I changed: 
1. I put the ME0 and GEM simlinks definition in the same package, SimDataFormats/GEMDigiSimLinks. 
2. I added the ME0 realistic digitizer from Roumyana Mileva (with associated python configurations)
3. I renamed the ME0 digi sequences to be clear and consistent. 
   real digis (`ME0Digi`) --> now `simMuonME0Digis`
   real pad digis (`ME0PadDigi`) --> now `simMuonME0PadDigis`
   fake digis (`ME0DigiPreReco`) --> now `simMuonME0PseudoDigis`
   fake re-digis (`ME0DigiPreReco`) --> now `simMuonME0PseudoReDigis`
   new ME0 digi sequence: `simMuonME0Digis + simMuonME0PadDigis + simMuonME0PseudoDigis + simMuonME0PseudoReDigis`
4. The new names were propagated all the way to L1, RECO and Validation.
5. I merged all customizations from `customizeGEMDigi`,`customizeME0Digi` and `customizeRPCDigi` in different packages to a single python file `customizeMuonDigi.py`. IMO there was way too much code-duplication in those customizations. I believe I merged all background customizations correctly. 
6. Background numbers for the legacy GEM and ME0 digitizers were updated. 
7. The `ME0PadDigiProducer` was updated to use the `ME0DigiCollection` instead of `ME0DigiPreRecoCollection`

Supporting information for the GEM and ME0 digitizers is being added here: https://twiki.cern.ch/twiki/bin/view/CMS/GEMDigitizationRoadMap

@calabria @mileva

Commit:
ME0 realistic digi updates

Commit:
ME0 realistic digi updates

Commit:
expert customs in customizeMuonDigi.py

Commit:
ME0 realistic digi updates

Commit:
ME0 realistic digi updates

Commit:
ME0 realistic digi updates

Conflicts:
	Validation/MuonME0Validation/python/MuonME0Digis_cfi.py

Commit:
new bkg model

Commit:
bugfix

Commit:
bugfix

Commit:
consistent naming

Commit:
update label

mileva committed:
background parameters fixed; obsolte code removed refference lumi param is taken in the configuration instead to be hardcoded; safety factor for beam pipe set from 2 to 1 since the new bkg model include it.

Commit:
fix identation

Commit:
delete obsolete file

Commit:
delete obsolete file

Commit:
fix typo

cmsbuild commented:
A new Pull Request was created by  @dildick (Sven Dildick) for master.

It involves the following packages:

IOMC/RandomEngine
RecoLocalMuon/GEMRecHit
RecoLocalMuon/GEMSegment
SimDataFormats/GEMDigiSimLink
SimMuon/Configuration
SimMuon/GEMDigitizer
SimMuon/RPCDigitizer
Validation/MuonME0Validation


@perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please review it and eventually sign? Thanks.
@calderona, @abbiendi, @jhgoh, @bellan, @HuguesBrun, @wddgit, @battibass, @trocino this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on RecoLocalMuon/GEMRecHit/python/me0RecHits_cfi.py:
```diff
@@ -3,5 +3,5 @@
     recAlgoConfig = cms.PSet(),
     recAlgo = cms.string('ME0RecHitStandardAlgo'),
-    me0DigiLabel = cms.InputTag("simMuonME0ReDigis"),
+    me0DigiLabel = cms.InputTag("simMuonME0PseudoReDigis"),
```
can we settle on some common naming so that (local) reco configs do not get perturbed frequently just because of some internal DIGI-step reorganization?

dildick replied:
@slava77 After this PR no more name changes are foreseen. The naming that is proposed is similar to the other muon digitizers `simMuonME0Digis` vs `simMuonRPCDigis` or `simMuonGEMDigis`,`simMuonCSCDigis` and `simMuonDTDigis` for realistic digis. `simMuonME0PadDigis` vs `simMuonGEMPadDigis` for pads. Pseudo digis now have the string `Pseudo` in them to distinguish them from real digis. Re-digitized pseudo digis for Muon TDR studies are called `simMuonME0PseudoReDigis`. This naming convention was agreed upon with other GEM/ME0 developers.

davidlange6 replied:
For my education, could you briefly summarize "pseudo" and "re"? (pseudo means smeared, and "re" means?) -perhaps the word "sim" needs to be dropped at some point along this chain

dildick replied:
Pseudo means that the digis are of type ME0DigiPreReco. Real digis are of type ME0Digi. The pseudo digitization sequence for ME0 was introduced for performance studies (deciding layers, eta partitions, strips) for the Muon Upgrade TDR. This sequence will be taken out of the main digi step in the future. The real ME0 digi sequence is what will be used to build ME0 triggers and for continued ME0 performance studies.

cmsbuild commented:
+1
Tested at: 17d06b5bde5f65ae47e495855e7281864c3ad7d1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19719/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/17d06b5bde5f65ae47e495855e7281864c3ad7d1
https://github.com/cms-sw/cmssw/commit/061366b2b0f5e9a53c988b709ec09a89cdb64557
https://github.com/cms-sw/cmssw/commit/7ec0bb687bcc23d9f4c72267916f6caddc77491b
https://github.com/cms-sw/cmssw/commit/c38f4427fd45907d68c6f37726b5d24371eb4332
https://github.com/cms-sw/cmssw/commit/385cde24ef413f269d4296778539fab588e7515e
https://github.com/cms-sw/cmssw/commit/47aa67168d028992f663bde7280aec7936d4d383
https://github.com/cms-sw/cmssw/commit/3fa70148e1221e5c6845d7f96b63a84c264756c0
https://github.com/cms-sw/cmssw/commit/634f58cfb9320b012ea61a9c2e5200fb96bf0ed5
https://github.com/cms-sw/cmssw/commit/b83376d6b29cd27b35386e543e94730bf4750e07
https://github.com/cms-sw/cmssw/commit/4710a08a52c987a42f6b3a98a225dd618ac77773
https://github.com/cms-sw/cmssw/commit/7c89cc40ad808792d856137ac245fef5576f743c
https://github.com/cms-sw/cmssw/commit/5fb33dd7fa3dc7f4f6b5edac213ce5c16f04bb1a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19719/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19719/git-merge-result

cmsbuild commented:
Comparison job queued.

slava77 commented:
On 5/10/17 11:57 AM, Sven Dildick wrote:
> *@dildick* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In RecoLocalMuon/GEMRecHit/python/me0RecHits_cfi.py
> <https://github.com/cms-sw/cmssw/pull/18660#discussion_r115824375>:
> 
>> @@ -3,5 +3,5 @@
>  me0RecHits = cms.EDProducer("ME0RecHitProducer",
>      recAlgoConfig = cms.PSet(),
>      recAlgo = cms.string('ME0RecHitStandardAlgo'),
> -    me0DigiLabel = cms.InputTag("simMuonME0ReDigis"),
> +    me0DigiLabel = cms.InputTag("simMuonME0PseudoReDigis"),
> 
> @slava77 <https://github.com/slava77> After this PR no more name changes
> are foreseen. The naming that is proposed is similar to the other muon
> digitizers |simMuonME0Digis| vs |simMuonRPCDigis| or
> |simMuonGEMDigis|,|simMuonCSCDigis| and |simMuonDTDigis| for realistic
> digis. |simMuonME0PadDigis| vs |simMuonGEMPadDigis| for pads. Pseudo
> digis now have the string |Pseudo| in them to distinguish them from real
> digis. Re-digitized pseudo digis for Muon TDR studies are called
> |simMuonME0PseudoReDigis|. This naming convention was agreed upon with
> other GEM/ME0 developers.

Is there a twiki or other well-established document that describes this?
The past few months looked a bit of a many-step walk that pushed product
name changes
to downstream applications for what's supposed to have the same physical
meaning
"digis useful for running standard reconstruction".


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/18660#discussion_r115824375>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbt0sLV4zrQewrmPcpMg6zr0h_iZRks5r4gi3gaJpZM4NWqbA>.
>

dildick commented:
We can put together some slides as a reference.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19719/summary.html

Comparison Summary:
 - You potentially added 19 lines to the logs
 - Reco comparison results: 1706 differences found in the comparisons
 - DQMHistoTests: Total files compared: 24
 - DQMHistoTests: Total histograms compared: 1829141
 - DQMHistoTests: Total failures: 55722
 - DQMHistoTests: Total nulls: 204
 - DQMHistoTests: Total successes: 1773035
 - DQMHistoTests: Total skipped: 180
 - DQMHistoTests: Total Missing objects: 0
 - Checked 98 log files, 14 edm output root files, 24 DQM output files

kpedro88 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,303 @@
+      {
+        const int time_hit(static_cast<int>(CLHEP::RandFlat::shoot(engine, nBxing)) + minBunch_);
+        std::pair<int, int> digi(k+1,time_hit);
+        strips_.insert(digi);
```
just use `strips_.emplace(k+1,time_hit)` to avoid the copy (also several instances further down)

kpedro88 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,303 @@
+    {
+      std::vector < std::pair<int, int> > cluster_;
+      cluster_.clear();
+      cluster_.push_back(std::pair<int, int>(centralStrip, time_hit));
```
use `emplace_back()` (also several instances further down)

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+  me0StripDigiSimLinkToken_(consumes<edm::DetSetVector<StripDigiSimLink> >(pset.getParameter<edm::InputTag>("me0StripDigiSimLinkToken")))
+  , debug_(pset.getParameter<bool>("debugFlag"))
+{
+  edm::Service < TFileService > fs; 
```
I think you need to have `usesResource("TFileService");` here

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+  edm::Service < TFileService > fs; 
+
+  hProces = fs->make < TH1F > ("hProces", "Process type for all the simHits", 20, 0, 20);
+//  hNstripEtaParts = fs->make <TH2F> ("NstripEtaParts", "Nstrips in each EtaPartition ", 40, 0.5, 10.5, 770, 1, 770);
```
delete commented-out code

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+
+void ME0DigiReader::analyze(const edm::Event & event, const edm::EventSetup& eventSetup)
+{
+//  cout << "--- Run: " << event.id().run() << " Event: " << event.id().event() << endl;
```
delete commented-out code

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+    const ME0DetId& id = (*detUnitIt).first;
+    const ME0EtaPartition* roll = pDD->etaPartition(id);
+
+    // ME0DetId print-out
```
delete commented-out code

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+    trStripArea = (roll->pitch()) * striplength;
+    trArea = trStripArea * nstrips;
+
+if(id.roll() == 1) { countRoll1++;}
```
indent these please

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+    for (ME0DigiCollection::const_iterator digiIt = range.first; digiIt!=range.second; ++digiIt)
+    {
+/*
+      for(int i = 0; i < id.roll(); i++){
```
delete commented-out code

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,368 @@
+std::cout << "roll 8 numbers = " <<  countRoll8 << "\tndigi = " << ndigi8 << std::endl;
+
+/*
+  for (edm::DetSetVector<StripDigiSimLink>::const_iterator itlink = thelinkDigis->begin(); itlink != thelinkDigis->end(); itlink++)
```
delete commented-out code

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc:
```diff
@@ -0,0 +1,459 @@
+  , me0DigiSimLinkToken_(consumes < edm::DetSetVector<ME0DigiSimLink> > (pset.getParameter < edm::InputTag > ("me0DigiSimLinkToken")))
+  , debug_(pset.getParameter<bool>("debugFlag"))
+{
+  edm::Service < TFileService > fs;
```
I think you need to have `usesResource("TFileService");` here

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc:
```diff
@@ -0,0 +1,459 @@
+
+    if (MuCluster.size() != 0)
+    {
+      for (std::map<int, int>::iterator it = MuCluster.begin(); it != MuCluster.end(); ++it)
```
use range-based loop (also below)

kpedro88 commented on SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc:
```diff
@@ -0,0 +1,459 @@
+          if (abs(particletype) == 13)
+          {
+            tof_mu_bx0->Fill(partTof);
+            MuCluster.insert(std::pair<int, int>(strip, particletype));
```
use `emplace()` (also below)

kpedro88 commented:
@dildick please only reply to my review comments on GitHub if you have a question or objection

dildick commented:
@kpedro88 ok.

@mileva could you please address the issues pertaining to the ME0 digi model? I'll take care of the muon digi customizations.

mileva commented:
@kpedro88  Thank you for the comments! 
@dildick  Ok, I will take care about the digi model and readers in the test directory, but can do this at the earliest on Saturday.

mileva commented on SimMuon/GEMDigitizer/src/ME0PreRecoGaussianModel.cc:
```diff
@@ -37,14 +37,20 @@ ME0PreRecoGaussianModel::ME0PreRecoGaussianModel(const edm::ParameterSet& config
-  neuBkg.push_back(-3.3405);      neuBkg.push_back(0.0140588);    neuBkg.push_back(-3.11473e-05); neuBkg.push_back(2.83736e-08);
-  eleBkg.push_back(4.68590e+05);  eleBkg.push_back(-1.63834e+04); eleBkg.push_back(2.35700e+02);
-  eleBkg.push_back(-1.77706e+00); eleBkg.push_back(7.39960e-03);  eleBkg.push_back(-1.61448e-05); eleBkg.push_back(1.44368e-08);
+  neuBkg.push_back(0.00386257);
```
It is better to be hard-coded. This was agreed previously to keep them here. The reason is that any little change in the parameters will lead to drastically wrong results and it is better to be not touched by mistake.

davidlange6 replied:
however, this approach does not lend itself to any sort of future proofing (eg, when your asked to evaluate a range of lumis..)

mileva commented on SimMuon/GEMDigitizer/src/GEMSimpleModel.cc:
```diff
@@ -34,42 +34,29 @@ GEMDigiModel(config)
-  GE21ModNeuBkgParam3 = -0.0185558;
-  GE21ModNeuBkgParam4 = 3.97809e-05;
-  GE21ModNeuBkgParam5 = -3.34575e-08;
+  //Electron Rate: model L=5x10^{34}cm^{-2}s^{-1}
```
It is better to be hard-coded. This was agreed previously to keep them here. The reason is that any little change in the parameters will lead to drastically wrong results and it is better to be not touched by mistake.

mileva commented on SimMuon/Configuration/python/customizeMuonDigi.py:
```diff
@@ -0,0 +1,353 @@
+        timeResolution = cms.double(2.5),
+        timingRPCOffset = cms.double(50.0)
+    ),
+    process.simMuonRPCReDigis.digiIRPCModelConfig = cms.PSet(
```
@bapavlov @jhgoh can you have a look on the RPC resolution constants here, please.

jhgoh replied:
Parameters in this customize function is expected to be modified "by hand" for a dedicated study. We want to keep the parameters to be visible to the analyzers.

jhgoh replied:
@dildick I made a PR to your branch to reduce repeated PSet values. Please have a look and merge it if you think it is OK.

jhgoh commented on SimMuon/RPCDigitizer/python/customizeRPCDigi.py:
```diff
@@ -1,152 +0,0 @@
-        process.rpcRecHits,
-        cms.Sequence(process.simMuonRPCReDigis+process.rpcRecHits)
-        )
-    return process
```
Are you removing this file? by accident?

dildick replied:
Not by accident. It is now included in SimMuon/Configuration/python/customizeMuonDigi.py which will contain all expert customizations for muon digi studies.

jhgoh replied:
OK. I understand.

slava77 commented:
For the names of digi collection used in reco, I'd rather it all had a name "simMuonME0Digis" as done for other detectors. It would decouple all internal reshuffling and renaming done in the DIGI step from downstream changes needed in RECO, DQM, Validation etc.

dildick commented:
While we are still using pseudo digis as input to the local ME0 reconstruction, the collection will be called "simMuonME0PseudoReDigis". Once an ME0 rechit builder is implemented that takes real digis "simMuonME0Digis" the ME0 system will be in line with the other detectors. All necessary changes downstream in RECO and Validation have been done in this PR (ME0 does not have a DQM module). If this motivation is still not satisfactory, could you suggest an alternative naming scheme?

slava77 commented:
On 5/15/17 12:22 PM, Sven Dildick wrote:
> While we are still using pseudo digis as input to the local ME0
> reconstruction, the collection will be called "simMuonME0PseudoReDigis".
> Once an ME0 rechit builder is implemented that takes real digis
> "simMuonME0Digis" the ME0 system will be in line with the other
> detectors. All necessary changes downstream in RECO and Validation have
> been done in this PR (ME0 does not have a DQM module). If this
> motivation is still not satisfactory, could you suggest an alternative
> naming scheme?

It looks like I'm missing some details.

My questions started with
  me0RecHits = cms.EDProducer("ME0RecHitProducer",
using simMuonME0PseudoReDigis as an input.

If I read your last comment correctly, you are saying that this rechit
producer is in some way incomplete
(as if it needs some pre-processed digis).

What will be the difference in the rechit builder that can consume
simMuonME0Digis?
When will it become available?


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/18660#issuecomment-301577218>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbpIPY_Lhm7vmxwpEnfnvnblMa1n6ks5r6KX8gaJpZM4NWqbA>.
>

dildick commented:
The current ME0 rechit builder takes the position and time of a single ME0 reprocessed pseudo-digi and makes an ME0 rechit out of it. The future ME0 rechit builder will be using real digis and will do digi clusterization similar as in the GEM rechit builder. Perhaps @calabria can provide more explanation. At the moment we do not have a time estimate for the realistic rechit builder.

Something I noticef is that I did not adjust the file "me0ChamberMasker_cff.py", although I doubt that is already used in sample production since "me0ChamberMasker_cfi" is missing in the release.

kpedro88 commented:
@dildick I think `me0ChamberMasker_cfi` might be auto-generated by the `fillDescriptions()` in `ME0ChamberMasker.cc`

dildick commented:
In that case the `me0ChamberMasker_cff` needs to be updated as well

Dr15Jones commented:
+1
Just a module added to the random number list

slava77 commented:
for reco
- I think the evolving "non-standard" input name for reco is acceptable, considering there is only one place to make this change now
- I'd like to understand changes visible in comparisons: in phase1/0 MC  RPCDigis plots with residuals are now empty. See example plot below. Is it intentional? (note that there are no changes in reco objects here, at least starting from RPC hits and downstream)

<img width="852" alt="wf10042_rpcdigi_simhitdiff" src="https://cloud.githubusercontent.com/assets/4676718/26130872/96a762d2-3a4b-11e7-897d-75deb2f37436.png">

slava77 commented:
@cmsbuild please test
to get reco comparisons for workflows missed in the last iteration of tests

cmsbuild commented:
The tests are being triggered in jenkins.

Commit:
use new ME0 digi module names

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jhgoh committed:
modify rpcRedigi customization function to be less-verbatim

jhgoh committed:
missing rpcdigi parameters

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
@mileva and I will address the remaining comments soon. You don't need to restart the tests.

cmsbuild commented:
+1
Tested at: cf50944a3777c833c68692434af499b683245f05
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19900/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/cf50944a3777c833c68692434af499b683245f05
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19900/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19900/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19900/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3369 differences found in the comparisons
 - DQMHistoTests: Total files compared: 24
 - DQMHistoTests: Total histograms compared: 1833858
 - DQMHistoTests: Total failures: 48761
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1784917
 - DQMHistoTests: Total skipped: 180
 - DQMHistoTests: Total Missing objects: 0
 - Checked 98 log files, 14 edm output root files, 24 DQM output files

Dr15Jones commented:
+1
for random number service

Commit:
address comments by Kevin Pedro

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
Still need to address the random number initialization

Commit:
load from IOMC

dildick commented on SimMuon/Configuration/python/customizeMuonDigi.py:
```diff
@@ -0,0 +1,353 @@
+
+
+# Add simMuonGEMDigis to the list of modules served by RandomNumberGeneratorService
+def customize_random_GEMDigi(process):
```
@kpedro88 If I use the construct 
<PRE>
# Add simMuonGEMDigis to the list of modules served by RandomNumberGeneratorService                                                                                                
def customize_random_GEMDigi(process):
    from IOMC.RandomEngine.IOMC_cff import RandomNumberGeneratorService as rndsv
    process.RandomNumberGeneratorService.simMuonGEMDigis = rndsv.simMuonGEMDigis
    return process
</PRE>

 in a driver command 

<PRE>
cmsDriver.py SingleMuPt100_pythia8_cfi \
--conditions auto:phase2_realistic -n 10 \
--era Phase2C2_timing \
--eventcontent FEVTDEBUG  \
-s GEN,SIM,DIGI:pdigi_valid \
--datatier GEN-SIM-DIGI \
--beamspot HLLHC \
--geometry Extended2023D4 \
--customise SimMuon/Configuration/customizeMuonDigi.customize_digi_addGEM_addME0_muon_only \
--python SingleMuPt100_2023tilted_GenSimFull.py \
--no_exec --fileout file:step1.root
</PRE>

then the driver crashes with 

<PRE>
...
    process = customize_random_GEMDigi(process)
  File "/afs/cern.ch/work/d/dildick/public/GEM/ME0DigiIntegration/CMSSW_9_2_X_2017-05-08-2300/python/SimMuon/Configuration/customizeMuonDigi.py", line 65, in customize_random_GEMDigi
    process.RandomNumberGeneratorService.simMuonGEMDigis = rndsv.simMuonGEMDigis
AttributeError: 'Service' object has no attribute 'simMuonGEMDigis'
</PRE>

How do I load the Phase2 `RandomNumberGeneratorService` before calling the customization?

dildick replied:
Or should I simply throw away the random number initialization and assume it will be done by the `--era` command?

kpedro88 replied:
Yes, you should assume the initialization will be done by the Era.

dildick replied:
ok.

Commit:
update configs

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
I removed the redundant random number initialization functions, updated the configs and ran some tests to make sure the new functions were correctly. A note was added at the top of `SimMuon/Configuration/python/customizeMuonDigi.py` which explains why those functions have disappeared.

jhgoh committed:
remove unnecessary parameter modifications by switching digi modelling

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d739aa90cbe55b4d9d1985587d2bc6bf8a6f8554
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19960/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
From @jhgoh: "I had chat with people working on RPC trigger and they confirmed the custom bunch-crossing spacing is not needed after switching the digi modeling."

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19960/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1690 differences found in the comparisons
 - DQMHistoTests: Total files compared: 24
 - DQMHistoTests: Total histograms compared: 1833969
 - DQMHistoTests: Total failures: 42897
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1790892
 - DQMHistoTests: Total skipped: 180
 - DQMHistoTests: Total Missing objects: 0
 - Checked 98 log files, 14 edm output root files, 24 DQM output files

milenadyu committed:
ME0 and GEMDigis Models are modified after KPedro comments

slava77 commented:
I see that a bunch of comments from @kpedro88  have not been addressed yet.
So, I'm not restarting the tests or checking for reco signoff yet.

dildick commented:
@slava77 We're working on it

cmsbuild commented:
Pull request #18660 was updated. @ghellwig, @lveldere, @kpedro88, @fabozzi, @Martin-Grunewald, @rekovic, @vanbesien, @govoni, @perrotta, @civanch, @monttj, @silviodonato, @cmsbuild, @fwyzard, @GurpreetSinghChahal, @davidlange6, @smuzaffar, @Dr15Jones, @kkousour, @mdhildreth, @cerminar, @slava77, @ggovi, @mmusich, @mulhearn, @ssekmen, @arunhep, @perrozzi, @thuer, @dmitrijus, @franzoni can you please check and sign again.

dildick commented:
This was not intended. I'll have to fix the branch.

kpedro88 commented:
-1
bad merge

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
@kpedro88 I'll address some remaining comments. Will let you know when I'm done.

Commit:
address comments by Kevin Pedro

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
Roumyana is working on moving the "magic numbers" into the python configs + adding explanation where they come from.

Commit:
address comments by Kevin Pedro

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

Commit:
address comments by Kevin Pedro

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
@kpedro88 I'm done. Please let me know if you spot other outdated C++ constructs.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b6e2af6b0bd32de341887b911c3aa61e674a3f04
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19976/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/19976/summary.html

Comparison Summary:
 - You potentially added 192 lines to the logs
 - Reco comparison results: 3362 differences found in the comparisons
 - DQMHistoTests: Total files compared: 24
 - DQMHistoTests: Total histograms compared: 1833935
 - DQMHistoTests: Total failures: 33905
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1799850
 - DQMHistoTests: Total skipped: 180
 - DQMHistoTests: Total Missing objects: 0
 - Checked 98 log files, 14 edm output root files, 24 DQM output files

milenadyu committed:
ME0 and GEM removed hardcoded BKG params

Conflicts:
	SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 53174fb7ae1d2e77cae7fe7fea5b7c10d310737c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20002/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20002/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 24
 - DQMHistoTests: Total histograms compared: 1833901
 - DQMHistoTests: Total failures: 18368
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1815353
 - DQMHistoTests: Total skipped: 180
 - DQMHistoTests: Total Missing objects: 0
 - Checked 98 log files, 14 edm output root files, 24 DQM output files

dildick commented:
@milenadyu @mileva Is there anything else you would like to add to this PR?

mileva commented:
@dildick No, for this PR we are not plannig something more.

dildick commented:
@kpedro88 Could you inspect one more time and sign off?

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/18660	53174fb
- reco is minimally affected: input tag changed for me0rechits
- jenkins tests pass and comparisons show no changes

kpedro88 commented:
+1

dildick commented:
Can this pull request be integrated, please? Thanks.

civanch commented:
+1

dmitrijus commented:
+1

calabria commented:
@davidlange6 @kpedro88 @slava77 is there anything else needed for this PR? if not can we proceed with the integration? This is really needed to start the me0 trigger studies.

Dr15Jones commented:
+1
for IOMC/RandomEngine

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

dildick commented:
@davidlange6 Can this pull request be integrated, please? Thank you.

davidlange6 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,286 @@
+    const int time_hit(static_cast<int>(CLHEP::RandFlat::shoot(engine, nBxing)) + minBunch_);
+    if (doNoiseCLS_)
+    {
+      std::vector < std::pair<int, int> > cluster_;
```
the trailing _ should be reserved for member data. (so remove it here)

davidlange6 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,286 @@
+    if (doNoiseCLS_)
+    {
+      std::vector < std::pair<int, int> > cluster_;
+      cluster_.clear();
```
no need to clear

davidlange6 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,286 @@
+  double deltaX = pointSimHit.x() - pointDigiHit.x();
+
+  // Add central digi to cluster vector
+  std::vector < std::pair<int, int> > cluster_;
```
same issues here.

davidlange6 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,286 @@
+  cluster_.emplace_back(centralStrip, bx);
+
+  //simulate cross talk
+  int clusterSize((CLHEP::RandFlat::shoot(engine)) <= 0.53 ? 1 : 2);
```
please document the 53%

milenadyu replied:
it is documented in https://twiki.cern.ch/twiki/bin/view/CMS/GEMDigitizationRoadMap

milenadyu replied:
"The probabilities of 53% for a single fired strip and 47% for two adjacent fired strips are chosen according to the first test results obtained in 2013 with additional tuning in order to have the cluster size of GEM RecHit relevant to the experimental measured ones."

davidlange6 commented on SimMuon/GEMDigitizer/src/ME0SimpleModel.cc:
```diff
@@ -0,0 +1,286 @@
+#include "CLHEP/Units/defs.h"
+#include "CLHEP/Units/SystemOfUnits.h"
+#include "CLHEP/Units/GlobalPhysicalConstants.h"
```
what requires these 3 includes?

kpedro88 replied:
`CLHEP::c_light` is used several times (at my request)

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,345 @@
+
+  ME0DigiCollection::DigiRangeIterator detUnitIt;
+
+  int countRoll1 = 0;
```
how about int countRoll[8]={0};

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,345 @@
+
+  TH1F *hProces; 
+
+  TH1F *hNstripEtaPart1;
```
similarly here - how about an array?

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,345 @@
+  int numbEvents;
+  int ndigi1, ndigi2, ndigi3, ndigi4, ndigi5, ndigi6, ndigi7, ndigi8;
+  double ndigiVsArea1, ndigiVsArea2, ndigiVsArea3, ndigiVsArea4, ndigiVsArea5, ndigiVsArea6, ndigiVsArea7, ndigiVsArea8;
+  double rollRadius1, rollRadius2, rollRadius3, rollRadius4, rollRadius5, rollRadius6, rollRadius7, rollRadius8;
```
and for these three

dildick commented:
@milenadyu and @mileva will address these issues.

milenadyu committed:
addressing questions by davidlange6 code optimizations (arrays)

milenadyu committed:
removing underscore from cluster-vector

cmsbuild commented:
Pull request #18660 was updated. @ghellwig, @ianna, @kpedro88, @fabozzi, @rekovic, @vanbesien, @perrotta, @civanch, @monttj, @cmsbuild, @GurpreetSinghChahal, @davidlange6, @smuzaffar, @Dr15Jones, @kkousour, @mdhildreth, @cerminar, @slava77, @ggovi, @mulhearn, @arunhep, @dmitrijus, @franzoni can you please check and sign again.

dildick commented:
I'll rebase on top of the latest build

kpedro88 commented:
-1
bad merge

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 53b942b4bdca35e7dec52bcf867ec3e5d7ec1fe6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20256/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/7965aa6b603fcb9a22e5142f3a1eb19236e0c7ca
https://github.com/cms-sw/cmssw/commit/c55feb9e2054fe6414f321a0a6ec9d6279d7121e
https://github.com/cms-sw/cmssw/commit/f3638a2c7e6d4fd1af2b1d13ed5c9c43535c1cc1
https://github.com/cms-sw/cmssw/commit/c6b7c566d21fd9dd0eb9b5bb19fc1cbc72b4ce28
https://github.com/cms-sw/cmssw/commit/bcc689f68972caa93024bf7bee8323d772a8d5ff
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20256/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20256/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20256/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18660/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D10_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D10+RecoFullGlobal_2023D10+HARVESTFullGlobal_2023D10
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18660/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D18_GenSimHLBeamSpotFull14+DigiFull_2023D18+RecoFullGlobal_2023D18+HARVESTFullGlobal_2023D18
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18660/27434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFull_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1688 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1786072
 - DQMHistoTests: Total failures: 38871
 - DQMHistoTests: Total nulls: 105
 - DQMHistoTests: Total successes: 1746923
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

dildick commented:
I believe we addressed all remaining issues.

kpedro88 commented:
+1

slava77 commented:
@cmsbuild please test

to get cleaner comparisons (last iteration of jenkins tests had more PRs merged which show some physics changes)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 53b942b4bdca35e7dec52bcf867ec3e5d7ec1fe6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20292/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20292/summary.html

Comparison Summary:
 - You potentially added 3 lines to the logs
 - Reco comparison results: 1669 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1789234
 - DQMHistoTests: Total failures: 37238
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1751823
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/18660	53b942b
- reco part were not changed since the last signoff for https://github.com/cms-sw/cmssw/pull/18660	53174fb
- jenkins tests pass and comparisons with baseline show no differences (including the 2023 workflows)

dildick commented:
Hi, can this PR now be approved? Thank you.

dmitrijus commented:
+1

mdhildreth commented:
+1

davidlange6 commented on SimDataFormats/GEMDigiSimLink/interface/ME0DigiSimLink.h:
```diff
@@ -0,0 +1,61 @@
+#ifndef ME0OBJECTS_ME0DIGISIMLINK_H
```
until ME0DigiSimLink diverges from GEMDigiSimLink, please just define a typedef.

dildick replied:
ME0DigiSimLink is now 
<PRE>
#include "SimDataFormats/GEMDigiSimLink/interface/GEMDigiSimLink.h"
typedef GEMDigiSimLink ME0DigiSimLink;
</PRE>

davidlange6 commented on SimMuon/GEMDigitizer/test/GEMDigiSimLinkReader.cc:
```diff
@@ -319,53 +319,23 @@ void GEMDigiSimLinkReader::analyze(const edm::Event & event, const edm::EventSet
+     // add electron and muon hits to cluster
+    for (const auto& p : MuCluster) myCluster.emplace(p);
+    for (const auto& p : ElecCluster) myCluster.emplace(p);
+    for (const auto& p : MuCluster)  if (abs(p.second) == 13) muonFired.emplace_back(p.first);
```
please use std::abs

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,315 @@
+    trStripArea = (roll->pitch()) * striplength;
+    trArea = trStripArea * nstrips;
+
+    if(id.roll() == 1) { countRoll[0]++;}
```
please use a loop

davidlange6 replied:
or rather if ( id.roll()>0 && id.roll()<9) countRoll[id.roll()-1]++;

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,315 @@
+    const ME0DigiCollection::Range& range = (*detUnitIt).second;
+    for (ME0DigiCollection::const_iterator digiIt = range.first; digiIt!=range.second; ++digiIt)
+    {
+      if(id.roll() == 1) {hNstripEtaParts->Fill(id.roll(),digiIt->strip()); ndigiEtaPart[0]++;}
```
similarly here.

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,315 @@
+
+      if (digiIt->strip() < 1 || digiIt->strip() > roll->nstrips() )
+      {
+        cout <<" XXXXXXXXXXXXX Problemt with "<<id<<"  a digi has strip# = "<<digiIt->strip()<<endl;
```
LogDebug?

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiReader.cc:
```diff
@@ -0,0 +1,315 @@
+
+  std::vector<double> myRadii, myRates;
+
+  std::cout << "ndigiVsArea1 = " << ndigiVsArea[0];
```
log debugs and loops

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc:
```diff
@@ -0,0 +1,224 @@
+      Muon_energy->Fill(simHit.pabs());
+    if (abs(simHit.particleType()) == 11)
+    {
+      if (simHit.processType() == 13)
```
else ifs or a case statement?

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc:
```diff
@@ -0,0 +1,224 @@
+    hProces->Fill(simHit.processType());
+    hAllSimHitsType->Fill(simHit.particleType());
+
+    if (abs(simHit.particleType()) == 13)
```
std::abs here

davidlange6 commented on SimMuon/GEMDigitizer/test/ME0DigiSimLinkReader.cc:
```diff
@@ -0,0 +1,224 @@
+        processtypeElectrons->Fill(processtype);
+      if (particletype == -11)
+        processtypePositrons->Fill(processtype);
+      if (abs(particletype) == 13)
```
std::abs

Commit:
suggestions by David Lange

cmsbuild commented:
Pull request #18660 was updated. @perrotta, @smuzaffar, @civanch, @Dr15Jones, @mdhildreth, @dmitrijus, @cmsbuild, @kpedro88, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

dildick commented:
I implemented the suggestions. Can you restart the tests?

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7098ea77e426bd1fd05f2c4b7139292e1d629b3d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20457/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18660/20457/summary.html

Comparison Summary:
 - You potentially added 5 lines to the logs
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1872323
 - DQMHistoTests: Total failures: 34634
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1837516
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/18660	7098ea7
- updates since last reco signoff were in the Sim\* subsystem, not related to reco https://github.com/cms-sw/cmssw/pull/18660/files/53b942b4bdca35e7dec52bcf867ec3e5d7ec1fe6..7098ea77e426bd1fd05f2c4b7139292e1d629b3d
- jenkins tests pass and comparisons show no differences

kpedro88 commented:
+1

dmitrijus commented:
+1

dildick commented:
@davidlange6 Do you have any more comments on this PR?

kpedro88 commented:
@Dr15Jones @civanch @mdhildreth please sign again

Dr15Jones commented:
+1
Just random number

mdhildreth commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @smuzaffar

dildick commented:
Can this pull request be signed please?

davidlange6 commented:
+1
